src/tree.rs

```diff
@@ -67,6 +67,64 @@ where
     }
 }
 
+pub fn join<'a, A>(tss: Tree<'a, Tree<'a, A>>) -> Tree<'a, A>
+where
+    A: Clone + 'a,
+{
+    bind(tss)(Rc::new(move |x| x))
+}
+
+pub fn duplicate<'a, A>(t: Tree<'a, A>) -> Tree<'a, Tree<'a, A>>
+where
+    A: Clone + 'a,
+{
+    let xs = t
+        .clone()
+        .children
+        .into_iter()
+        .map(|x| duplicate(x))
+        .collect();
+    Tree::new(t, xs)
+}
+
+pub fn fold<A, X, B, F, G>(f: Rc<F>) -> impl Fn(Rc<G>) -> Rc<dyn Fn(Tree<A>) -> B>
+where
+    A: Clone,
+    B: Clone,
+    X: Clone,
+    // TODO get rid of these static lifetimes
+    F: Fn(A) -> Rc<dyn Fn(X) -> B> + 'static,
```

Comment:
this is weird. I can't seem to get the lifetimes to line up as they normally would (by supplying `'a` or alternatively).

As such, I'm kind of just going to get this to compile and leave the above note and try to solve this later as I think there might be a nice way to get these to do what we want using the iterator adapter stuff so things play more idiomatically with rust's semantics around ownership and borrowing. That basically means writing an `Iterator` impl for `Tree`; you get the rest for free once you have an instance of an Iterator, which could be the same type or a different one.